install_config/persistent_storage/persistent_storage_csi.adoc

```diff
@@ -0,0 +1,412 @@
+[[install-config-persistent-storage-persistent-storage-csi]]
+= Persistent Storage Using Container Storage Interface (CSI)
+{product-author}
+{product-version}
+:data-uri:
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+:prewrap!:
+
+toc::[]
+
+== Overview
+Container Storage Interface (CSI) allows {product-title} to consume storage from storage backends that implement the link:https://github.com/container-storage-interface/spec[CSI interface] as xref:../../architecture/additional_concepts/storage.adoc#architecture-additional-concepts-storage[persistent
+storage].
+
+
+[NOTE]
+====
+CSI volumes is an alpha feature and may change in a future release of {product-title}.
+====
+
+[NOTE]
+====
+{product-title} does not ship with any CSI drivers. It is recommended to use the CSI drivers provided by link:https://kubernetes-csi.github.io/docs/Drivers.html[community or storage vendors].
+====
+
+[NOTE]
+====
+{product-title} {product-version} supports version 0.2.0 of link:https://github.com/container-storage-interface/spec[CSI specification].
+====
+
+
+[[install-config-persistent-storage-csi-architecture]]
+== Architecture
+
+CSI drivers are typically shipped as container images. These containers are not aware of {product-title} where they run. To use CSI-compatible storage backed in {product-title}, the cluster administrator must deploy several components that serve as a bridge between {product-title} and the storage driver.
+
+The following diagram provides high-level overview about these components running in pods in {product-title} cluster.
+
+image::csi.png["Architecture of CSI components"]
+
+It is possible to run multiple CSI drivers for different storage backends. Each driver needs its own external attacher StatefulSet, external provisioner deployment, and DaemonSet with the driver and CSI registrar.
```

Comment:
In openshift 3.10, a cluster can have more than one csi drivers?

Reply:
> What different between csi provisioner and csi attacher makes we using StatefulSet to create attacher pod(replica=1) and Deployment to create provisioner pod(replica=2)?

@qinpingli, the attached does not support any leader election so far, so only one attacher can run at a time. And StatefulSet with replica count=1 ensures that. Deployment with replica count=1 can run two pods in some corner cases.

This will be fixed in 3.11 and attacher will run as Deployment there. Actually, the patch has just been merged, so I could port it back to 3.10 and simplify deployment and docs. Do you still have time to re-test it?

Reply:
and yes, it should be possible to run more CSI drivers in 3.10.

Reply:
Cool! I'd like to re-test it.